Implement cross-language await syntax for KJ Promises and Rust Futures

harrishancock opened the pull request:
Associated pull request(s):
- https://github.com/capnproto/capnproto/pull/2216
- https://github.com/capnproto/capnproto/pull/2220
- https://github.com/capnproto/capnproto/pull/2223

harrishancock committed:
Implement cross-language await syntax for KJ Promises and Rust Futures

harrishancock committed:
[nfc] Improve rust/async testing

harrishancock committed:
Factor out unsafe code

danlapid commented:
Would it be reasonable to request a README.md on the src/rust/async directory explaining what we're doing, the approach and maybe pointers as to how to understand the code? Which part interacts with which and in what way?
Was trying to read it and while I have some context it's hard to grok how everything interacts 😢

harrishancock commented:
Definitely! I won't ask for review until code comments are fleshed out, and I have an explainer written.

harrishancock committed:
Implement foreign Waker case

harrishancock committed:
Improve safety with ExecutorGuarded and LinkedGroup/LinkedObject

This commit also includes various other refactors.

harrishancock committed:
Refactoring, renaming, async tracing

jasnell commented:
The c++ parts are looking good. One concern I do have is this is a fairly large complicated bit of code... would it make sense / be possible for it to be split out into a separate dependency outside of workerd?

harrishancock commented:
@jasnell Yes, I agree. I think the capnproto repo is the most logical home for most of this, maybe in kj/compat. Or a separate repo, call it kj-rs, could also make sense.

I'll keep it here in draft form for now while I iterate, but will think about how to support this in capnproto's build.

harrishancock committed:
Update FutureAwaiter implementation to new capnproto branch

harrishancock committed:
Use Arc::disown()/reown()

harrishancock committed:
wip exception propagation

harrishancock committed:
Fix typo

harrishancock committed:
Add LinkedObjectList::empty()

harrishancock committed:
Refactor in preparation for exception propagation

harrishancock committed:
Adjust comment for accuracy

Awaiter was renamed to PromiseAwaiter in a recent capnproto update.

harrishancock committed:
Update capnproto dep

harrishancock committed:
Simplify ArcWaker error path

This seemed like unjustifiable complexity.

harrishancock committed:
Rename awaitBegin() -> suspend()

harrishancock committed:
Flesh out comments

harrishancock committed:
Future<Result<()>> can throw to C++ when awaited

harrishancock committed:
Rename box_future_... forwarding functions

harrishancock committed:
Await Promise<void> instead of OwnPromiseNode

harrishancock committed:
Promise<T> can return Err Result to Rust when awaited

harrishancock committed:
Refactor boilerplate to reduce coupling

harrishancock committed:
Promise<T> can return T to Rust when awaited

harrishancock committed:
BoxFuture<T> can return T to C++ when awaited

harrishancock committed:
Pass actual Waker reference to C++

harrishancock committed:
Remove poll() overloads taking CoAwaitWaker

harrishancock committed:
Refactor for clarity

CoAwaitWaker is now a simpler data structure, with a new FuturePollEvent base class taking on the burden of connecting the various events and promises.

harrishancock committed:
Get rid of extra setDone()/isDone() API

harrishancock committed:
Rename KjWaker -> LazyArcWaker, simplify API

It turns out we can return a Maybe<Promise<void>> from LazyArcWaker::reset() with our desired semantics.

harrishancock committed:
Rename CxxWaker -> KjWaker and add tryGetFuturePollEvent() virtual function

This allows us to avoid a dynamic_cast to implement our optimization.  Having a separate CxxWaker base class is a bit purer, but doesn't seem worth the trouble.

harrishancock committed:
Simplify CoAwaitWaker

harrishancock committed:
Move CoAwaitWaker from await.h/c++ to waker.h/c++

harrishancock committed:
Move WakerRef, OptionWaker into await_.rs

harrishancock committed:
Rename await.* awaiter.*

The await_.rs bit was bugging me.

harrishancock committed:
Move Future boilerplate into future-boilerplate.h/c++

harrishancock committed:
Reorganize boilerplate

harrishancock committed:
Simplify BoxFutureFulfiller<T>

harrishancock committed:
Make tracing work more like ExclusiveJoinPromiseNode

harrishancock committed:
Rename CoAwaitWaker -> FuturePollEvent::PollScope

harrishancock committed:
Move Promise boilerplate

harrishancock committed:
Remove `workerd::async::rust::Promise<T>` alias template

It occurred to me that it is not necessary, since we have to make type
aliases for each specific type anyway.

harrishancock committed:
Start a README.md

harrishancock committed:
Add missing include
